jupyterhub/crypto.py

```diff
@@ -0,0 +1,161 @@
+
+import base64
+from binascii import a2b_hex
+from concurrent.futures import ThreadPoolExecutor
+import json
+import os
+
+from traitlets.config import SingletonConfigurable, Config
+from traitlets import (
+    Any, Dict, Integer, List,
+    default, observe, validate,
+)
+
+try:
+    import cryptography
+    from cryptography.fernet import Fernet, MultiFernet, InvalidToken
+except ImportError:
+    cryptography = None
+    class InvalidToken(Exception):
+        pass
+
+
+KEY_ENV = 'JUPYTERHUB_CRYPT_KEY'
+
+class EncryptionUnavailable(Exception):
+    pass
+
+class CryptographyUnavailable(EncryptionUnavailable):
+    def __str__(self):
+        return "cryptography library is required for encryption"
+
+class NoEncryptionKeys(EncryptionUnavailable):
+    def __str__(self):
+        return "Encryption keys must be specified in %s env" % KEY_ENV
+
+
+def _validate_key(key):
+    """Validate and return a 32B key
+
+    Args:
+    key (bytes): The key to be validated.
+        Can be:
+        - base64-encoded (44 bytes)
+        - hex-encoded (64 bytes)
+        - raw 32 byte key
+
+    Returns:
+    key (bytes): raw 32B key
+    """
+    if isinstance(key, str):
+        key = key.encode('ascii')
+
+    if len(key) == 44:
+        try:
+            key = base64.urlsafe_b64decode(key)
+        except ValueError:
+            pass
+
+    elif len(key) == 64:
+        try:
+            # 64B could be 32B, hex-encoded
+            return a2b_hex(key)
+        except ValueError:
+            # not 32B hex
+            pass
+
+    if len(key) != 32:
+        raise ValueError("Encryption keys must be 32 bytes, hex or base64-encoded.")
+
+    return key
+
+class CryptKeeper(SingletonConfigurable):
+    """Encapsulate encryption configuration
+
+    Use via the encryption_config singleton below.
+    """
+
+    n_threads = Integer(max(os.cpu_count(), 1), config=True,
```

Comment:
How many threads you should use depends on what the threads are doing. Since encryption is CPU-bound, it shouldn't be greater than n cpus, or we'll just thrash context switching as the actual CPU threads hop between the large number of threads we have asked it to run concurrently that it can't actually do. Other pools are waiting on I/O, so a larger number makes sense, but concurrent connection limits (for example) can also dictate an upper bound for those that could easily be lower than the number of concurrent spawns allowed. Each spawn goes through many stages, and one thread waiting on an HTTP response doesn't conflict with another spawn from encrypting its auth_state, so it's not nearly as simple as setting a single concurrency limit and applying it to all resources.

> For example, if n_threaads is 1 we're effectively limiting the number of concurrent user logins to 1

This isn't true. Limiting the number of users that can be in a *single stage* of login isn't the same as limiting the number of concurrent logins, especially when one value causes a user-facing error message, while the other causes a minor increase in login time. Pool size is setting a bound on the amount of CPU/memory resources a given task can be allowed to ask for at a given time. Encryption is designed to be CPU and memory intensive, so allowing 1000 encryption tasks to spawn at the same time doesn't make sense on an 8 core server.